src/Node/BlockNode.php

```diff
@@ -37,14 +38,7 @@ public function compile(Compiler $compiler): void
 
         $compiler
             ->subcompile($this->getNode('body'))
-        ;
-
-        if (!$this->getNode('body') instanceof NodeOutputInterface && $compiler->getEnvironment()->useYield()) {
-            // needed when body doesn't yield anything
-            $compiler->write("yield '';\n");
-        }
-
-        $compiler
+            ->write("return; yield '';\n") // needed when body doesn't yield anything
```

Comment:
Not all nodes that do generate some content have the `NodeOutputInterface`. This does the trick of forcing an empty generator without any overhead.

Reply:
nodes generating output without implementing NodeOutputInterface should also be deprecated though, as some other parts of Twig rely on this interface. Not sure how to handle that.

Reply:
I don't see a way to force node implementations to have this interface if they want to output something. The only side-effect of implementing the interface is forbidding output-enabled nodes from being use outside of blocks, so it's not a critical feature either.